pkg/kubelet/server.go

```diff
@@ -587,80 +595,160 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
 Loop:
 	for {
 		select {
 		case <-conn.CloseChan():
+			glog.V(5).Infof("Upgraded connection closed")
 			break Loop
 		case stream := <-streamChan:
-			streamType := stream.Headers().Get(api.StreamType)
-			port := stream.Headers().Get(api.PortHeader)
-			dataStreamLock.Lock()
-			switch streamType {
-			case "error":
-				ch := make(chan httpstream.Stream)
-				dataStreamChans[port] = ch
-				go waitForPortForwardDataStreamAndRun(kubecontainer.GetPodFullName(pod), uid, stream, ch, s.host)
-			case "data":
-				ch, ok := dataStreamChans[port]
-				if ok {
-					ch <- stream
-					delete(dataStreamChans, port)
-				} else {
-					glog.Errorf("Unable to locate data stream channel for port %s", port)
-				}
-			default:
-				glog.Errorf("streamType header must be 'error' or 'data', got: '%s'", streamType)
-				stream.Reset()
+			connectionID := getPortForwardStreamConnectionID(stream)
+			glog.V(5).Infof("(%s) Received new stream", connectionID)
+
+			streamLock.Lock()
+
+			ch, ok := streamChans[connectionID]
+			if !ok {
+				glog.V(5).Infof("(%s) connection ID not in streamChans", connectionID)
+
+				glog.V(5).Infof("(%s) creating stream channel", connectionID)
+				ch = make(chan httpstream.Stream)
+				streamChans[connectionID] = ch
+
+				glog.V(5).Infof("(%s) starting handlePortForwardStreams", connectionID)
+				go handlePortForwardStreams(podName, uid, ch, portForwarder, connectionID, streamChans, streamLock)
 			}
-			dataStreamLock.Unlock()
+
+			glog.V(5).Infof("(%s) sending stream to stream channel", connectionID)
+			ch <- stream
+
+			streamLock.Unlock()
 		}
 	}
 }
 
-func waitForPortForwardDataStreamAndRun(pod string, uid types.UID, errorStream httpstream.Stream, dataStreamChan chan httpstream.Stream, host HostInterface) {
-	defer errorStream.Reset()
+// handlePortForwardStreams waits to receive the error and data streams via streamChan,
+// then uses portForwarder to perform the actual port forwarding logic. If the streams
+// are not received within streamCreationTimeout, this function returns.
+func handlePortForwardStreams(pod string, uid types.UID, streamChan chan httpstream.Stream, portForwarder PortForwarder, connectionID string, streamChans map[string]chan httpstream.Stream, streamChansLock sync.Mutex) {
+	// remove streamChan when forwarding is done
+	defer func() {
+		streamChansLock.Lock()
+		defer streamChansLock.Unlock()
+		delete(streamChans, connectionID)
+		close(streamChan)
+	}()
 
-	var dataStream httpstream.Stream
+	var errorStream, dataStream httpstream.Stream
 
-	select {
-	case dataStream = <-dataStreamChan:
-	case <-time.After(streamCreationTimeout):
-		errorStream.Write([]byte("Timed out waiting for data stream"))
-		//TODO delete from dataStreamChans[port]
-		return
+	for {
+		select {
+		case stream := <-streamChan:
+			defer stream.Close()
+
+			streamType := stream.Headers().Get(api.StreamType)
+			glog.V(5).Infof("(%s) handlePortForwardStreams got stream type %s", connectionID, streamType)
```

Comment:
Nit: I think this message could be a little cleared wrt `connectionID`.  How about:

`"handlePortForwardStreams get stream type %s for connectionID %s"`